# Author: Chenyuan Li

In [ ]:
import pandas as pd
import numpy as np
import math
import os

## Task 1

In [9]:
MSA1_prop_stat = pd.read_excel("MSA1.xlsx", sheet_name="Property Status", header = 4)

In [7]:
MSA2_prop_stat = pd.read_excel("MSA2.xlsx", sheet_name="Property Status", header = 4)

In [12]:
MSA1_prop_stat.head(5)

MarketCode             MarketName StateCode StateName  CountyCode  \
0       12420  Austin-Round Rock, TX        TX     Texas         453   
1       12420  Austin-Round Rock, TX        TX     Texas         453   
2       12420  Austin-Round Rock, TX        TX     Texas         209   
3       12420  Austin-Round Rock, TX        TX     Texas         453   
4       12420  Austin-Round Rock, TX        TX     Texas         453   

  CountyName TractCode  BlockGroupCode  Latitude  Longitude  ...  Dec-18  \
0     Travis    000305             2.0  30.30917  -97.72881  ...       S   
1     Travis    001716             2.0  30.36687  -97.79303  ...       S   
2       Hays    010400             3.0  29.84896  -97.94931  ...      LU   
3     Travis    001860             2.0  30.45009  -97.64873  ...       S   
4     Travis    000604             3.0  30.28462  -97.74723  ...       S   

  Jan-19 Feb-19  Mar-19 Apr-19 May-19 Jun-19 Jul-19 Aug-19  Sep-19  
0      S      S       S      S      S      S      S      S       S  
1      S      S       S      S      S      S      S      S       S  
2     LU     LU       S      S      S      S      S      S       S  
3      S      S       S      S      S      S      S      S       S  
4      S      S       S      S      S      S      S      S       S  

[5 rows x 168 columns]

In [11]:
MSA2_prop_stat.head()

MarketCode MarketName StateCode StateName  CountyCode CountyName  \
0       10420  Akron, OH        OH      Ohio         153     Summit   
1       10420  Akron, OH        OH      Ohio         153     Summit   
2       10420  Akron, OH        OH      Ohio         153     Summit   
3       10420  Akron, OH        OH      Ohio         153     Summit   
4       10420  Akron, OH        OH      Ohio         153     Summit   

   TractCode  BlockGroupCode  Latitude  Longitude  ...  Dec-18 Jan-19 Feb-19  \
0   530104.0             4.0  41.31810  -81.44502  ...       S      S      S   
1   530603.0             1.0  41.17920  -81.48354  ...       S      S      S   
2   530401.0             1.0  41.17894  -81.43911  ...       S      S      S   
3        NaN             NaN  41.10510  -81.46834  ...       S      S      S   
4   507500.0             1.0  41.13701  -81.54175  ...       S      S      S   

   Mar-19 Apr-19 May-19 Jun-19 Jul-19 Aug-19  Sep-19  
0       S      S      S      S      S      S       S  
1       S      S      S      S      S      S       S  
2       S      S      S      S      S      S       S  
3       S      S      S      S      S      S       S  
4       S      S      S      S      S      S       S  

[5 rows x 168 columns]

In [14]:
def count_delivered(df):
    # initialize count
    count = 0 
    # create empty dictionary, key is the row number of the property, value is the column number
    # when that property is delivered (will use in next part)
    delivered_month = {}
    for i in range(len(df)):
        for j in range(list(df.columns).index("Apr-08"), len(list(df.columns))):
            if type(df.iloc[i,j]) == str:
                if df.iloc[i,j] == 'LU':
                    count += 1
                    delivered_month[i] = j
                    break
                elif df.iloc[i,j] == "UC/LU":
                    count += 1
                    delivered_month[i] = j
                    break
                else:
                    break
            elif type(df.iloc[i, j]) == float:
                continue
    return count, delivered_month
            
count_MSA1, delivered_month_MSA1 = count_delivered(MSA1_prop_stat)
count_MSA2, delivered_month_MSA2 = count_delivered(MSA2_prop_stat)
print(count_MSA1)
print(count_MSA2)

270
9


### The number of properties delivered is 270 for Austin-Round Rock, TX and 9 for Akron, OH since April 2008.

## Calculate average lease-up time

In [16]:
MSA1_occ = pd.read_excel("MSA1.xlsx", sheet_name="Occ & Concession", header = 4)
MSA2_occ = pd.read_excel("MSA2.xlsx", sheet_name="Occ & Concession", header = 4)

In [17]:
MSA1_occ.head()

MarketCode             MarketName StateCode StateName  CountyCode  \
0       12420  Austin-Round Rock, TX        TX     Texas         453   
1       12420  Austin-Round Rock, TX        TX     Texas         453   
2       12420  Austin-Round Rock, TX        TX     Texas         209   
3       12420  Austin-Round Rock, TX        TX     Texas         453   
4       12420  Austin-Round Rock, TX        TX     Texas         453   

  CountyName TractCode  BlockGroupCode  Latitude  Longitude  ...  Dec-18.2  \
0     Travis    000305             2.0  30.30917  -97.72881  ...   0.00000   
1     Travis    001716             2.0  30.36687  -97.79303  ...   0.00000   
2       Hays    010400             3.0  29.84896  -97.94931  ...   0.06934   
3     Travis    001860             2.0  30.45009  -97.64873  ...   0.00000   
4     Travis    000604             3.0  30.28462  -97.74723  ...   0.00000   

   Jan-19.2  Feb-19.2  Mar-19.2 Apr-19.2  May-19.2  Jun-19.2  Jul-19.2  \
0  0.000000  0.000000  0.000000  0.00000  0.000000  0.000000  0.000000   
1  0.000000  0.000000  0.000000  0.00000  0.052353  0.041026  0.000000   
2  0.083054  0.083054  0.083054  0.06914  0.069140  0.010221  0.035623   
3  0.000000  0.000000  0.000000  0.00000  0.000000  0.000000  0.000000   
4  0.000000  0.000000  0.000000  0.00000  0.000000  0.000000  0.000000   

  Aug-19.2  Sep-19.2  
0      0.0       0.0  
1      0.0       0.0  
2      0.0       0.0  
3      0.0       0.0  
4      0.0       0.0  

[5 rows x 444 columns]

In [18]:
MSA2_occ.head()

MarketCode MarketName StateCode StateName  CountyCode CountyName  \
0       10420  Akron, OH        OH      Ohio         153     Summit   
1       10420  Akron, OH        OH      Ohio         153     Summit   
2       10420  Akron, OH        OH      Ohio         153     Summit   
3       10420  Akron, OH        OH      Ohio         153     Summit   
4       10420  Akron, OH        OH      Ohio         153     Summit   

   TractCode  BlockGroupCode  Latitude  Longitude  ...  Dec-18.2 Jan-19.2  \
0   530104.0             4.0  41.31810  -81.44502  ...       0.0      0.0   
1   530603.0             1.0  41.17920  -81.48354  ...       0.0      0.0   
2   530401.0             1.0  41.17894  -81.43911  ...       0.0      0.0   
3        NaN             NaN  41.10510  -81.46834  ...       0.0      0.0   
4   507500.0             1.0  41.13701  -81.54175  ...       0.0      0.0   

   Feb-19.2  Mar-19.2 Apr-19.2 May-19.2 Jun-19.2 Jul-19.2 Aug-19.2  Sep-19.2  
0  0.000000       0.0      0.0      0.0      0.0      0.0      0.0       0.0  
1  0.000000       0.0      0.0      0.0      0.0      0.0      0.0       0.0  
2  0.100728       0.0      0.0      0.0      0.0      0.0      0.0       0.0  
3  0.000000       0.0      0.0      0.0      0.0      0.0      0.0       0.0  
4  0.000000       0.0      0.0      0.0      0.0      0.0      0.0       0.0  

[5 rows x 444 columns]

In [19]:
# Find where occupancy columns end, filter out other columns
list(MSA1_occ.columns).index("Apr-08.1")

168

In [20]:
MSA1_occ = MSA1_occ.iloc[:,0:168]
MSA2_occ = MSA2_occ.iloc[:,0:168]

In [22]:
def lease_up(df, delivered_month):
    # create empty list for each property's lease-up time
    month = []
    for k,v in delivered_month.items():
        for j in range(list(df.columns).index("Apr-08"), len(list(df.columns))):
            if df.iloc[k, j] >= 0.9: # find first month its occupancy rate reaches 90% or above
                month.append(j - v) # find difference between column number 
                break
            else:
                continue
    return np.mean(month)

avg_month_MSA1 = lease_up(MSA1_occ, delivered_month_MSA1)
avg_month_MSA2 = lease_up(MSA2_occ, delivered_month_MSA2)
print(avg_month_MSA1)
print(avg_month_MSA2)


12.721311475409836
8.75


### The average lease-up time is 12.72 months for Austin-Round Rock, TX and 8.75 months for Akron, OH.

## Task 2.1

In [23]:
# create an empty dataframe, set columns, and append all states' dataframes to this dataframe
df = pd.DataFrame(columns =['State_in', 'County_in', 'State_from', 'County_from', 'State', 'County',
                                             'Number of Returns', 'Number of Exemptions', 'Adjusted Gross Income'])
for file in os.listdir():
    if file.endswith('xls'):
        state = pd.read_excel(file)
        state.columns = ['State_in', 'County_in', 'State_from', 'County_from', 'State', 'County',
                                             'Number of Returns', 'Number of Exemptions', 'Adjusted Gross Income']
        state = state.iloc[7:,:]
        df = df.append(state, ignore_index = True)

In [27]:
df

State_in County_in State_from County_from State  \
0            10       000         96         000    DE   
1            10       000         97         000    DE   
2            10       000         97         001    DE   
3            10       000         97         003    DE   
4            10       000         98         000    DE   
...         ...       ...        ...         ...   ...   
106410       47       189         59         000    DS   
106411       47       189         59         001    DS   
106412       47       189         59         003    DS   
106413       47       189         59         005    DS   
106414       47       189         59         007    DS   

                          County Number of Returns Number of Exemptions  \
0           Total Mig - US & For             18621                34538   
1                 Total Mig - US             18149                33744   
2         Total Mig - US Same St              3325                 6212   
3         Total Mig - US Diff St             14824                27532   
4            Total Mig - Foreign               472                  794   
...                          ...               ...                  ...   
106410  Other Flows - Diff State               813                 1721   
106411   Other Flows - Northeast                69                  162   
106412     Other Flows - Midwest               171                  346   
106413       Other Flows - South               462                  948   
106414        Other Flows - West               111                  265   

       Adjusted Gross Income  
0                     849929  
1                     836221  
2                     128324  
3                     707896  
4                      13709  
...                      ...  
106410                 36349  
106411                  3355  
106412                  7599  
106413                 21498  
106414                  3897  

[106415 rows x 9 columns]

In [29]:
# found error entries, drop them
df = df[df['Number of Returns'] != 'd60']
df = df[df['Number of Returns'] != 'r 81']

In [30]:
# convert specific column types to integers
df = df.astype({'State_in': 'int64', 'State_from': 'int64', 'Number of Returns': 'int64',
               'Number of Exemptions': 'int64', 'Adjusted Gross Income': 'int64'})

In [32]:
# Filter out total migration rows by selecting State_from < 55, and non-migrants (when state_in = state_from and 
# County_in = County_from)

# Find maximum number of exemptions
np.max(df[(df['State_in'] != df['State_from']) & (df['County_in'] != df['County_from']) & 
   (df['State_from'] < 55)]['Number of Exemptions'])

11955

In [33]:
# Find which county is it from
df[df['Number of Exemptions'] == 11955]

State_in County_in  State_from County_from State              County  \
47170        32       003           6         037    CA  Los Angeles County   

       Number of Returns  Number of Exemptions  Adjusted Gross Income  
47170               5624                 11955                 210568

In [34]:
# Find whhich county is the County_in in the above row
df[(df['State_from'] == 32) & (df['County_from'] == '003')].head()

State_in County_in  State_from County_from State        County  \
186         10       003          32         003    NV  Clark County   
799         30       013          32         003    NV  Clark County   
906         30       029          32         003    NV  Clark County   
966         30       031          32         003    NV  Clark County   
1132        30       049          32         003    NV  Clark County   

      Number of Returns  Number of Exemptions  Adjusted Gross Income  
186                  13                    28                   1003  
799                  20                    35                    808  
906                  26                    46                    788  
966                  25                    57                   1242  
1132                 15                    26                    477

### Clark County in NV and Los Angeles County in CA have the most interactions.

## Task 2.2

In [36]:
# Get most in-migration county by using the State_from = 96 and County_from == 000, but County_in != 000
np.max(df[(df['State_from'] == 96) & (df['County_in'] != '000') & (df['County_from'] == '000')]['Number of Exemptions'])

170749

In [37]:
df[df['Number of Exemptions'] == 170749]

State_in County_in  State_from County_from State  \
98290         6       037          96         000    CA   

                                 County  Number of Returns  \
98290  Los Angeles Cou Tot Mig-US & For             102342   

       Number of Exemptions  Adjusted Gross Income  
98290                170749                4151171

### Most in-migration takes place in Los Angeles County in CA.

## Task 2.3

In [38]:
# Group by State_from and County_from, find the sum of Number of Exemptions, use the same filter
# for Task 2.1
df[(df['State_from'] < 55) & (df['State_in'] != df['State_from']) 
   & (df['County_in'] != df['County_from'])].groupby(['State_from', 
                                                      'County_from'])['Number of Exemptions'].sum().idxmax()

(6, '037')

### The above state and county codes indicate Los Angeles County in CA have the largest out-migration.

## Task 2.4

In [40]:
# All counties in NY metro area
NY_metro = [
'Bergen County',
'Dutchess County',
'Essex County',
'Fairfield County',
'Hudson County'
'Hunterdon County',
'Litchfield County',
'Mercer County',
'Middlesex County',
'Monmouth County',
'Morris County',
'Nassau County',
'New Haven County',
'Ocean County',
'Orange County',
'Passaic County',
'Pike County',
'Putnam County',
'Rockland County',
'Somerset County',
'Suffolk County',
'Sussex County',
'Ulster County',
'Union County',
'Westchester County']

In [41]:
# Find total in-migrations into NY metro counties
total_in = 0
for county in NY_metro:
    county_column = county + ' Tot Mig-US & For'
    if df[df['County'] == county_column]['Number of Exemptions'].empty:
        total_in = total_in
    else:
        total_in += df[df['County'] == county_column]['Number of Exemptions'].iloc[0]
    
print(total_in)

229551


In [42]:
# Find total out-migrations from NY metro counties
total_out = 0
for county in NY_metro:
    if df[df['County'] == county].empty:
        total_out = total_out
    else:
        for i in range(len(df[df['County'] == county])):
            total_out += df[df['County'] == county]['Number of Exemptions'].iloc[i]
print(total_out)

805658


In [44]:
net_migration = total_in - total_out
print(net_migration)

-576107


### Net migration for NY metropolitan area is -576107.

## Task 2.5

The challenging part was to think about how to deal with 50 separate excel files. I solved it by using os.listdr() to read all files and import them into an empty dataframe. All the other tasks became much more straightforward because one can do filtering, subsetting, and use group-by clauses to find specific answer to each task.